File: shared_resources/src/main/java/com/thewizrd/shared_resources/database/LocationsDAO.java
package com.thewizrd.shared_resources.database;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.thewizrd.shared_resources.locationdata.LocationData;
import com.thewizrd.shared_resources.weatherdata.Favorites;

import java.util.List;

@Dao
public interface LocationsDAO {
    /* LocationData methods */
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public void insertLocationData(LocationData Location);

    @Update
    public void updateLocationData(LocationData location);

    @Delete
    public void deleteLocationData(LocationData Location);

    @Query("DELETE FROM locations")
    public void deleteAllLocationData();

    @Query("DELETE FROM locations WHERE `query` = :key")
    public void deleteLocationDataByKey(String key);

    @Transaction
    @Query("SELECT * FROM locations")
    public List<LocationData> loadAllLocationData();

    @Transaction
    @Query("SELECT locations.`query`, locations.`name`, locations.`latitude`," +
            " locations.`longitude`, locations.`tz_long`, locations.`locationType`, locations.`source`, locations.`locsource` FROM locations" +
            " INNER JOIN favorites ON locations.`query` = favorites.`query`" +
            " ORDER BY favorites.position")
    public List<LocationData> getFavorites();

    @Query("SELECT * FROM locations WHERE `query` = :query")
    public LocationData getLocation(String query);

    @Query("SELECT COUNT(*) FROM locations")
    public int getLocationDataCount();

    /* Favorites methods */
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public void insertFavorite(Favorites favorite);

    @Update
    public void updateFavorite(Favorites favorite);

    @Delete
    public void deleteFavoriteData(Favorites favorite);

    @Query("DELETE FROM favorites")
    public void deleteAllFavoriteData();

    @Query("DELETE FROM favorites WHERE `query` = :key")
    public void deleteFavoritesByKey(String key);

    @Transaction
    @Query("SELECT * FROM favorites")
    public List<Favorites> loadAllFavorites();

    @Query("SELECT * FROM favorites WHERE `query` = :query")
    public List<Favorites> getFavoritesData(String query);

    @Transaction
    @Query("SELECT * FROM favorites ORDER BY `position`")
    public List<Favorites> loadAllFavoritesByPosition();

    @Query("SELECT COUNT(*) FROM favorites")
    public int getFavoritesCount();

    @Query("UPDATE favorites SET `position` = :toPos WHERE `query` = :key")
    public void updateFavPosition(String key, int toPos);
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/database/LocationsDAO.kt
package com.thewizrd.shared_resources.database

import androidx.room.*
import com.thewizrd.shared_resources.locationdata.LocationData
import com.thewizrd.shared_resources.weatherdata.model.Favorites

@Dao
interface LocationsDAO {
    /* LocationData methods */
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertLocationData(Location: LocationData)

    @Update
    suspend fun updateLocationData(location: LocationData)

    @Delete
    suspend fun deleteLocationData(Location: LocationData)

    @Query("DELETE FROM locations")
    suspend fun deleteAllLocationData()

    @Query("DELETE FROM locations WHERE `query` = :key")
    suspend fun deleteLocationDataByKey(key: String)

    @Transaction
    @Query("SELECT * FROM locations")
    suspend fun loadAllLocationData(): List<LocationData>

    @Query(
        "SELECT locations.* FROM locations" +
                " INNER JOIN favorites ON locations.`query` = favorites.`query`" +
                " ORDER BY favorites.position"
    )
    @Transaction
    suspend fun getFavorites(): List<LocationData>

    @Query(
        "SELECT locations.* FROM locations" +
                " INNER JOIN favorites ON locations.`query` = favorites.`query`" +
                " ORDER BY favorites.position LIMIT 1"
    )
    @Transaction
    suspend fun getFirstFavorite(): LocationData?

    @Query("SELECT * FROM locations WHERE `query` = :query")
    suspend fun getLocation(query: String?): LocationData?

    @Query("SELECT COUNT(*) FROM locations")
    suspend fun getLocationDataCount(): Int

    @Query("SELECT COUNT(*) FROM locations")
    fun getLocationDataCountSync(): Int

    /* Favorites methods */
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertFavorite(favorite: Favorites)

    @Update
    suspend fun updateFavorite(favorite: Favorites)

    @Delete
    suspend fun deleteFavoriteData(favorite: Favorites)

    @Query("DELETE FROM favorites")
    suspend fun deleteAllFavoriteData()

    @Query("DELETE FROM favorites WHERE `query` = :key")
    suspend fun deleteFavoritesByKey(key: String?)

    @Transaction
    @Query("SELECT * FROM favorites")
    suspend fun loadAllFavorites(): List<Favorites>

    @Query("SELECT * FROM favorites WHERE `query` = :query LIMIT 1")
    suspend fun getFavorite(query: String?): Favorites?

    @Transaction
    @Query("SELECT * FROM favorites ORDER BY `position`")
    suspend fun loadAllFavoritesByPosition(): List<Favorites>

    @Query("SELECT COUNT(*) FROM favorites")
    suspend fun getFavoritesCount(): Int

    @Query("UPDATE favorites SET `position` = :toPos WHERE `query` = :key")
    suspend fun updateFavPosition(key: String?, toPos: Int)
}